dali/dfu/dfuutil.cpp

```diff
@@ -857,21 +857,56 @@ class CDFUhelper: public CInterface, implements IDFUhelper
             else
                 throwError1(DFUERR_DFileNotFound, lfn);
         }
-        else {
-            Owned<IFileDescriptor> fdesc = file->getFileDescriptor();
-            Owned<IPropertyTree> t = fdesc->getFileTree();
+        else
+        {
+            Owned<IPropertyTree> t = queryDistributedFileDirectory().getFileTree(lfn, user);
             toXML(t, out, true);
         }
         return out;
     }
 
-    void addFileXML(const char *lfn,const StringBuffer &xml,IUserDescriptor *user)
+    void addFileXML(const char *lfn, const StringBuffer &xml, IUserDescriptor *user)
     {
         Owned<IPropertyTree> t = createPTreeFromXMLString(xml);
-        Owned<IFileDescriptor> fdesc = deserializeFileDescriptorTree(t,&queryNamedGroupStore(),0);
-        Owned<IDistributedFile> file = queryDistributedFileDirectory().createNew(fdesc,true);
-        if (file)
-            file->attach(lfn,user);
+        IDistributedFileDirectory &dfd = queryDistributedFileDirectory();
+
+        if (dfd.exists(lfn, user))
+            throw MakeStringException(-1, "Destination file '%s' already exists!", lfn);
+
+        // Check if this XML is a superfile map
+        const char * nodeName = t->queryName();
+        if (0 == strcmp(nodeName, queryDfsXmlBranchName(DXB_SuperFile)))
+        {
+            // It seems XML is fine to create  super file
+            Owned<IDistributedSuperFile> superfile = dfd.createNewSuperFile(t);
+
+            superfile->validate();
+
+            try
+            {
+                superfile->attach(lfn, user, true);
+            }
+            catch (IException *e)
+            {
+                EXCLOG(e,"addFileXML(Super File");
+                e->Release();
+            }
+        }
+        else if (0 == strcmp(nodeName, queryDfsXmlBranchName(DXB_File)))
+        {
+            // Logical file map
+
+            Owned<IFileDescriptor> fdesc = deserializeFileDescriptorTree(t, &queryNamedGroupStore(), 0);
+            Owned<IDistributedFile> file = dfd.createNew(fdesc, true);
+            if (file)
+            {
+                file->validate();
+                PROGLOG("Add logical file %s.", lfn);
+                file->attach(lfn,user);
+            }
```

Comment:
With validate() and attach() being common virtuals (see previous comment re. single version of attach() prototype), and other comment re. exception handling, this code could be simplified to:

```
        Owned<IDistributedFile> file;
        if (0 == strcmp(nodeName, queryDfsXmlBranchName(DXB_SuperFile)))
        {
            // It seems XML is a super file
            file.setown(dfd.createNewSuperFile(t));
        }
        else if (0 == strcmp(nodeName, queryDfsXmlBranchName(DXB_File)))
        {
            // Logical file map
            Owned<IFileDescriptor> fdesc = deserializeFileDescriptorTree(t, &queryNamedGroupStore(), 0);
            file.setown(dfd.createNew(fdesc, true));
        }
        file->validate();
        PROGLOG("Adding %s file %s.", file->querySuperFile()?"super":"logical", lfn);
        file->attach(lfn,user);
```